x-pack/elastic-agent/pkg/core/server/server.go

```diff
@@ -156,9 +156,13 @@ func (s *Server) Start() error {
 
 	// start serving GRPC connections
 	go func() {
+		if s.server == nil { // Server.Stop was called before this goroutine run
+			s.logger.Error("cannot start gRPC server, server is nil. Did you call Stop before the initialization has completed?")
+			return
+		}
```

Comment:
I don't see how this check could help.
What `Stop` function you're talking about, this `Server` struct does not have any? And even if it did, we would need to synchronise access to `s.server` with a lock. Otherwise, something can set `s.server` to `nil` after this check and before `s.server.Serve`.

Reply:
yes, the fix ins't perfect. In the e2e tests it seems this server is being closed and therefore `s.server` being set to `nil` before this goroutine runs. 

I don't remember all the cases when the scheduler would stop the goroutine, but I'm betting it's less likely that the scheduler would delay to run the goroutine, then it would stop this giroutine between the if and the actual call to `s.server.Serve`.

Reply:
So, the [`Stop`function](https://github.com/elastic/beats/blob/0e067a23b8ea4876d63c6e184e31fabaa7eb58ed/x-pack/elastic-agent/pkg/core/server/server.go#L182) sets `s.server` to nil. And looking at the logs on the tests failing because a panic when [starting the server](https://github.com/elastic/beats/blob/0e067a23b8ea4876d63c6e184e31fabaa7eb58ed/x-pack/elastic-agent/pkg/core/server/server.go#L163), the culprit seems to be the fact the agent starts and stops the server immediately. Therefore it's possible, and logical, to believe this goroutine is running after `Stop`, thus the `s.server` is `nil`.

And as I said, the check will help to avoid the panic happening, even thought it won't prevent it 100%.

Reply:
I see but I still believe we need a lock for the whole `Stop` section and this code would become something like this.

```suggestion
		s.startLock.Lock()
		server := s.server
		s.startLock.Unlock()
		if server == nil { // Server.Stop was called before this goroutine run
			s.logger.Error("cannot start gRPC server, server is nil. Did you call Stop before the initialization has completed?")
			return
		}
```